README.md
![SendGrid Logo](twilio_sendgrid_logo.png)

[![Test and Deploy](https://github.com/sendgrid/smtpapi-nodejs/actions/workflows/test-and-deploy.yml/badge.svg)](https://github.com/sendgrid/smtpapi-nodejs/actions/workflows/test-and-deploy.yml)
[![npm](https://img.shields.io/npm/v/smtpapi.svg)](https://www.npmjs.com/package/smtpapi)
[![MIT licensed](https://img.shields.io/badge/license-MIT-blue.svg)](LICENSE)
[![Twitter Follow](https://img.shields.io/twitter/follow/sendgrid.svg?style=social&label=Follow)](https://twitter.com/sendgrid)
[![GitHub contributors](https://img.shields.io/github/contributors/sendgrid/smtpapi-nodejs.svg)](https://github.com/sendgrid/smtpapi-nodejs/graphs/contributors)

**This node module allows you to quickly and more easily generate SendGrid X-SMTPAPI headers.**

# Announcements
**The default branch name for this repository has been changed to `main` as of 07/27/2020.**

All updates to this module are documented in our [CHANGELOG](CHANGELOG.md).

# Table of Contents
- [Installation](#installation)
- [Quick Start](#quick-start)
- [Library Usage Documentation](USAGE.md)
- [How to Contribute](#contribute)
- [About](#about)
- [License](#license)

<a name="installation"></a>
# Installation

## Prerequisites

- Node.js version 4, 5, 6, 7, 8, 10, 12, 14, or 16
- The SendGrid service, starting at the [free level](https://sendgrid.com/free?source=smtpapi-nodejs)

## Environment Variables

Update your environment with your SendGrid credentials. (if you will be sending via nodemailer or other SMTP transport)

```bash
echo "export SENDGRID_USERNAME='YOUR_USERNAME'" > sendgrid.env
echo "export SENDGRID_PASSWORD='YOUR_PASSWORD'" >> sendgrid.env
echo "sendgrid.env" >> .gitignore
source ./sendgrid.env
```

The following recommended installation requires [npm](https://npmjs.org/). If you are unfamiliar with npm, see the [npm docs](https://npmjs.org/doc/). Npm comes installed with Node.js since node version 0.8.x, therefore, you likely already have it.

Install smtpapi-nodejs and its dependencies:

```bash
npm install smtpapi
```

## Dependencies

- Please see [package.json](package.json)

<a name="quick-start"></a>
# Quick Start

```javascript
var smtpapi = require("smtpapi");
var header = new smtpapi();
header.addTo("test@example.com");
header.setUniqueArgs({ cow: "chicken" });
console.log(header.jsonString());
```

<a name="usage"></a>
# Usage

<a name="contribute"></a>
# How to Contribute

We encourage contribution to our projects, please see our [CONTRIBUTING](CONTRIBUTING.md) guide for details.

Quick links:

- [Feature Request](CONTRIBUTING.md#feature-request)
- [Bug Reports](CONTRIBUTING.md#submit-a-bug-report)
- [Improvements to the Codebase](CONTRIBUTING.md#improvements-to-the-codebase)
- [Review Pull Requests](CONTRIBUTING.md#code-reviews)

<a name="about"></a>
# About

smtpapi-nodejs is maintained and funded by Twilio SendGrid, Inc. The names and logos for smtpapi-nodejs are trademarks of Twilio SendGrid, Inc.

If you need help installing or using the library, please check the [Twilio SendGrid Support Help Center](https://support.sendgrid.com).

If you've instead found a bug in the library or would like new features added, go ahead and open issues or pull requests against this repo!

<a name="license"></a>
# License
[The MIT License (MIT)](LICENSE)


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.

[2022-03-09] Version 1.4.7
--------------------------
**Library - Chore**
- [PR #113](https://github.com/sendgrid/smtpapi-nodejs/pull/113): push Datadog Release Metric upon deploy success. Thanks to [@eshanholtz](https://github.com/eshanholtz)!


[2022-02-09] Version 1.4.6
--------------------------
**Library - Chore**
- [PR #112](https://github.com/sendgrid/smtpapi-nodejs/pull/112): upgrade supported language versions. Thanks to [@childish-sambino](https://github.com/childish-sambino)!
- [PR #111](https://github.com/sendgrid/smtpapi-nodejs/pull/111): add gh release to workflow. Thanks to [@shwetha-manvinkurke](https://github.com/shwetha-manvinkurke)!


[2022-01-12] Version 1.4.5
--------------------------
**Library - Chore**
- [PR #110](https://github.com/sendgrid/smtpapi-nodejs/pull/110): update license year. Thanks to [@JenniferMah](https://github.com/JenniferMah)!
- [PR #109](https://github.com/sendgrid/smtpapi-nodejs/pull/109): migrate to gh actions. Thanks to [@shwetha-manvinkurke](https://github.com/shwetha-manvinkurke)!


[2020-09-02] Version 1.4.4
--------------------------
**Library - Docs**
- [PR #84](https://github.com/sendgrid/smtpapi-nodejs/pull/84): Ran .md files through Grammer.ly. Thanks to [@adamk1230](https://github.com/adamk1230)!


[2020-08-19] Version 1.4.3
--------------------------
**Library - Chore**
- [PR #107](https://github.com/sendgrid/smtpapi-nodejs/pull/107): update GitHub branch references to use HEAD. Thanks to [@thinkingserious](https://github.com/thinkingserious)!


[2020-02-19] Version 1.4.2
--------------------------
**Library - Test**
- [PR #99](https://github.com/sendgrid/smtpapi-nodejs/pull/99): Update unittest removing unnecessary scenarios. Thanks to [@rafaelverger](https://github.com/rafaelverger)!


[2020-01-22] Version 1.4.1
--------------------------
**Library - Docs**
- [PR #105](https://github.com/sendgrid/smtpapi-nodejs/pull/105): baseline all the templated markdown docs. Thanks to [@childish-sambino](https://github.com/childish-sambino)!


[2020-01-15] Version 1.4.0
--------------------------
**Library - Docs**
- [PR #35](https://github.com/sendgrid/smtpapi-nodejs/pull/35): add table of contents in README.md. Thanks to [@thepriefy](https://github.com/thepriefy)!
- [PR #33](https://github.com/sendgrid/smtpapi-nodejs/pull/33): Making README/Doc Sections More SEO Friendly. Thanks to [@mgingras](https://github.com/mgingras)!
- [PR #39](https://github.com/sendgrid/smtpapi-nodejs/pull/39): fix typo and ToC in contributing and readme docs. Thanks to [@pushkyn](https://github.com/pushkyn)!
- [PR #44](https://github.com/sendgrid/smtpapi-nodejs/pull/44): Add Pull Request template file. Thanks to [@jeffgerlach](https://github.com/jeffgerlach)!
- [PR #52](https://github.com/sendgrid/smtpapi-nodejs/pull/52): add usage.md. Thanks to [@rohitdaryanani](https://github.com/rohitdaryanani)!
- [PR #34](https://github.com/sendgrid/smtpapi-nodejs/pull/34): Added Troubleshooting.md. Thanks to [@akram-rameez](https://github.com/akram-rameez)!
- [PR #64](https://github.com/sendgrid/smtpapi-nodejs/pull/64): Added TROUBLESHOOTING.md. Thanks to [@dave4506](https://github.com/dave4506)!
- [PR #49](https://github.com/sendgrid/smtpapi-nodejs/pull/49): Add a .env_sample file and update gitignore. Thanks to [@thepriefy](https://github.com/thepriefy)!
- [PR #40](https://github.com/sendgrid/smtpapi-nodejs/pull/40): add GitHub ISSUE_TEMPLATE. Thanks to [@thepriefy](https://github.com/thepriefy)!
- [PR #75](https://github.com/sendgrid/smtpapi-nodejs/pull/75): Added Code Review to Contributing.md. Thanks to [@mptap](https://github.com/mptap)!
- [PR #74](https://github.com/sendgrid/smtpapi-nodejs/pull/74): Add code reviews to CONTRIBUTING.md. Thanks to [@dagjaneiro](https://github.com/dagjaneiro)!
- [PR #71](https://github.com/sendgrid/smtpapi-nodejs/pull/71): make Environmental Variables a sub topic in CONTRIBUTING.md. Thanks to [@thepriefy](https://github.com/thepriefy)!
- [PR #80](https://github.com/sendgrid/smtpapi-nodejs/pull/80): Update LICENSE.txt year. Thanks to [@nocategory](https://github.com/nocategory)!
- [PR #97](https://github.com/sendgrid/smtpapi-nodejs/pull/97): Add our Developer Experience Engineer career opportunity to the README. Thanks to [@mptap](https://github.com/mptap)!
- [PR #103](https://github.com/sendgrid/smtpapi-nodejs/pull/103): just testing ... sorry. Thanks to [@thinkingserious](https://github.com/thinkingserious)!
- [PR #101](https://github.com/sendgrid/smtpapi-nodejs/pull/101): nothing to see here. Thanks to [@thinkingserious](https://github.com/thinkingserious)!
- [PR #102](https://github.com/sendgrid/smtpapi-nodejs/pull/102): add newline to README. Thanks to [@thinkingserious](https://github.com/thinkingserious)!

**Library - Fix**
- [PR #37](https://github.com/sendgrid/smtpapi-nodejs/pull/37): Update travis config node_js versions. Thanks to [@gr8shivam](https://github.com/gr8shivam)!
- [PR #38](https://github.com/sendgrid/smtpapi-nodejs/pull/38): Freeze dev dependencies so tests pass on 10.x. Thanks to [@mgingras](https://github.com/mgingras)!
- [PR #66](https://github.com/sendgrid/smtpapi-nodejs/pull/66): updating to fit codeclimate's new format/params. Thanks to [@mbernier](https://github.com/mbernier)!

**Library - Test**
- [PR #60](https://github.com/sendgrid/smtpapi-nodejs/pull/60): Added unittest to check for specific repo files. Thanks to [@mptap](https://github.com/mptap)!
- [PR #62](https://github.com/sendgrid/smtpapi-nodejs/pull/62): Add a unittest to check the license.md file date range. Thanks to [@lvarayut](https://github.com/lvarayut)!

**Library - Chore**
- [PR #63](https://github.com/sendgrid/smtpapi-nodejs/pull/63): Update the license file to have the correct date range. Thanks to [@lvarayut](https://github.com/lvarayut)!

test/main.js
var assert      = require('assert');
var should      = require('should');
var smtpapi     = require('../lib/main');
var t           = require('./smtpapi_test_strings.json');
var fs          = require('fs');

describe('smtapi', function() {
  beforeEach(function() {
    header = new smtpapi();
  });

  it('has a jsonString method', function() {
    header.jsonString().should.eql(t.json_string);
  });

  it('addTo', function() {
    header.addTo('addTo@mailinator.com');
    header.jsonString().should.eql(t.add_to);
  });

  it('setTos', function() {
    header.setTos(['setTos@mailinator.com']);
    header.jsonString().should.eql(t.set_tos);
  });

  it('addSubstitution', function() {
    header.addSubstitution('sub', 'val');
    header.jsonString().should.eql(t.add_substitution);
  });

  it('addSubstitution array value', function() {
    header.setSubstitutions({'sub': ['val 1']});
    header.addSubstitution('sub', ['val 2']);
    header.jsonString().should.eql(t.add_substitution_array_value);
  });

  it('setSubstitutions', function() {
    header.setSubstitutions({'sub': ['val']});
    header.jsonString().should.eql(t.set_substitutions);
  });

  it('addUniqueArg', function() {
    header.addUniqueArg('add_unique_argument_key', 'add_unique_argument_value');
    header.addUniqueArg('add_unique_argument_key_2', 'add_unique_argument_value_2');
    header.jsonString().should.eql(t.add_unique_arg);
  });

  it('setUniqueArgs', function() {
    header.setUniqueArgs({set_unique_argument_key: 'set_unique_argument_value'});
    header.jsonString().should.eql(t.set_unique_args);
  });

  it('addCategory', function() {
    header.addCategory('addCategory');
    header.addCategory('addCategory2');
    header.jsonString().should.eql(t.add_category);
  });

  it('addCategory', function() {
    header.addCategory(['addCategory']);
    header.jsonString().should.equal(t.add_category_array);
  });

  it('addCategoryUnicode', function() {
    header.addCategory('カテゴリUnicode');
    header.addCategory('カテゴリ2Unicode');
    header.jsonString().should.eql(t.add_category_unicode);
  });

  it('setCategories', function() {
    header.setCategories(['setCategories']);
    header.jsonString().should.eql(t.set_categories);
  });

  it('addSection', function() {
    header.addSection('set_section_key', 'set_section_value');
    header.addSection('set_section_key_2', 'set_section_value_2');
    header.jsonString().should.eql(t.add_section);
  });

  it('setSections', function() {
    header.setSections({'set_section_key': 'set_section_value'});
    header.jsonString().should.eql(t.set_sections);
  });

  it('addFilter', function() {
    header.addFilter('footer', 'text/html', '<strong>boo</strong>');
    header.jsonString().should.eql(t.add_filter);
  });

  it('setFilters', function() {
    var filter = {
      'footer': {
        'setting': {
          'enable': 1,
          'text/plain': 'You can haz footers!'
        }
      }
    };

    header.setFilters(filter);
    header.jsonString().should.eql(t.set_filters);
  });

  it('setSendAt', function() {
    header.setSendAt(1409348513);
    header.jsonString().should.eql(t.set_send_at);
  });

  it('setBatchId', function() {
    header.setBatchId('apples');
    header.jsonString().should.eql(t.set_batch_id);
  });

  it('setSendAt', function() {
    header.setSendEachAt([1409348513, 1409348514, 1409348515]);
    header.jsonString().should.eql(t.set_send_each_at);
  });

  it('addSendEachAt', function() {
    header.addSendEachAt(1409348513);
    header.addSendEachAt(1409348514);
    header.addSendEachAt(1409348515);
    header.jsonString().should.eql(t.add_send_each_at);
  });

  it('setSendEachAt and addSendEachAt', function() {
    header.setSendEachAt([1409348513]);
    header.addSendEachAt(1409348514);
    header.addSendEachAt(1409348515);
    header.jsonString().should.eql(t.add_send_each_at);
  });

  it('setASMGroupID', function() {
    header.setASMGroupID(123);
    header.jsonString().should.eql(t.set_asm_group_id);
  });

  it('setIpPool', function() {
    header.setIpPool('testPool');
    header.jsonString().should.eql(t.set_ip_pool);
  });
});

describe('smtapi-nodejs repo', function() {
  it('should have ./.env_sample file', function() {
    assert(fileExists('.env_sample'));
  });

  it('should have ./.gitignore file', function() {
    assert(fileExists('.gitignore'));
  });

  it('should have ./.github/workflows/test-and-deploy.yml file', function() {
    assert(fileExists('.github/workflows/test-and-deploy.yml'));
  });

  it('should have ./.codeclimate.yml file', function() {
    assert(fileExists('.codeclimate.yml'));
  });

  it('should have ./CHANGELOG.md file', function() {
    assert(fileExists('CHANGELOG.md'));
  });

  it('should have ./CODE_OF_CONDUCT.md file', function() {
    assert(fileExists('CODE_OF_CONDUCT.md'));
  });

  it('should have ./CONTRIBUTING.md file', function() {
    assert(fileExists('CONTRIBUTING.md'));
  });

  it('should have ./ISSUE_TEMPLATE.md file', function() {
    assert(fileExists('ISSUE_TEMPLATE.md'));
  });

  it('should have ./LICENSE file', function() {
    assert(fileExists('LICENSE'));
  });

  it('should have ./PULL_REQUEST_TEMPLATE.md file', function() {
    assert(fileExists('PULL_REQUEST_TEMPLATE.md'));
  });

  it('should have ./README.md file', function() {
    assert(fileExists('README.md'));
  });

  it('should have ./TROUBLESHOOTING.md file', function() {
    assert(fileExists('TROUBLESHOOTING.md'));
  });

  it('should have ./USAGE.md file', function() {
    assert(fileExists('USAGE.md'));
  });

  function fileExists(filepath) {
    try {
      return fs.statSync(filepath).isFile();
    } catch(e) {
      if (e.code === 'ENOENT') {
        console.log('' + filepath + ' doesn\'t exist.');
        return false;
      }
      throw e;
    }
  }
});


test/license.js
'use strict';

/**
 * Dependencies
 */
const fs = require("fs");
const assert = require('assert');

/**
 * Tests
 */
describe('LICENSE', () => {
  it('should contain the current year as the end year of the license', () => {
    const license = fs.readFileSync(`${process.cwd()}/LICENSE`, 'utf8');
    const currentYear= (new Date()).getFullYear();
    return assert(license.indexOf(`${currentYear}`) !== -1);
  });
});


lib/main.js
(function() {
"use strict";

var package_json  = require('./../package.json');

var smtpapi = function smtpapi(header) {
  this.version = package_json.version;

  header = header || {};
  this.header = {};
  this.header.to = header.to || [];
  this.header.sub = header.sub || {};
  this.header.unique_args = header.unique_args || {};
  this.header.category = header.category || [];
  this.header.section = header.section || {};
  this.header.filters = header.filters || {};
  this.header.send_at = header.send_at || '';
  this.header.batch_id = header.batch_id || '';
  this.header.send_each_at = header.send_each_at || [];
  this.header.asm_group_id = header.asm_group_id || {};
  this.header.ip_pool = header.ip_pool || '';
};

smtpapi.prototype.addTo = function(to) {
  if (to instanceof Array) {
    this.header.to = this.header.to.concat(to);
  } else {
    this.header.to.push(to);
  }
};

smtpapi.prototype.setTos = function(to) {
  if (to instanceof Array) {
    this.header.to = to;
  } else {
    this.header.to = [to];
  }
};

smtpapi.prototype.addSubstitution = function(key, val) {
  if (this.header.sub[key] === undefined) this.header.sub[key] = [];
  if (val instanceof Array) {
    this.header.sub[key] = this.header.sub[key].concat(val);
  } else {
    this.header.sub[key].push(val);
  }
};

smtpapi.prototype.setSubstitutions = function(subs) {
  this.header.sub = subs;
};

smtpapi.prototype.addUniqueArg = function(key, val) {
  this.header.unique_args[key] = val;
};

smtpapi.prototype.setUniqueArgs = function(val) {
  this.header.unique_args = val;
};

smtpapi.prototype.addCategory = function(cat) {
  if (cat instanceof Array) {
    this.header.category = this.header.category.concat(cat);
  } else {
    this.header.category.push(cat);
  }
};

smtpapi.prototype.setCategories = function(cats) {
  if (cats instanceof Array) {
    this.header.category = cats;
  } else {
    this.header.category = [cats];
  }
};

smtpapi.prototype.addSection = function(sec, val) {
  this.header.section[sec] = val;
};

smtpapi.prototype.setSections = function(sec) {
  this.header.section = sec;
};

smtpapi.prototype.addFilter = function(filter, setting, val) {
  if (this.header.filters[filter] === undefined) {
    this.header.filters[filter] = {'settings': {}};
  }
  this.header.filters[filter].settings[setting] = val;
};

smtpapi.prototype.setFilters = function(filters) {
  this.header.filters = filters;
};

smtpapi.prototype.setSendAt = function(send_at) {
  this.header.send_each_at = [];
  this.header.send_at = send_at;
};

smtpapi.prototype.setBatchId = function(batch_id) {
  this.header.batch_id = batch_id;
};

smtpapi.prototype.setSendEachAt = function(send_each_at) {
  this.header.send_at = '';
  this.header.send_each_at = send_each_at
};

smtpapi.prototype.addSendEachAt = function(send_each_at) {
  this.header.send_at = '';
  this.header.send_each_at.push(send_each_at);
};

smtpapi.prototype.setASMGroupID = function(asm_group_id) {
  this.header.asm_group_id = asm_group_id;
};

smtpapi.prototype.setIpPool = function(ip_pool) {
  this.header.ip_pool = ip_pool;
};

smtpapi.prototype.jsonObject = function() {
  var header = {};
  for (var key in this.header) {
    var isObject = (typeof this.header[key] === 'object' && Object.keys(this.header[key]).length);
    var headerHasKey = this.header[key]
    var isNumber = typeof this.header[key] === 'number' && headerHasKey;
    var isString = typeof this.header[key] === 'string' && headerHasKey;
    if (this.header.hasOwnProperty(key) && isObject ||
                                           isNumber ||
                                           isString) {
      header[key] = this.header[key];
    }
  }

  return header;
};

smtpapi.prototype.jsonString = function() {
  var json_object = this.jsonObject();
  var json_string = JSON.stringify(json_object);

  return this.escapeUnicode(json_string);
};

smtpapi.prototype.escapeUnicode = function(str) {
  return str.replace(/[^ -~]/g, function(m0) {
    var code = m0.charCodeAt(0);
    return '\\u' + ((code < 0x10)? '000' :
                    (code < 0x100)? '00' :
                    (code < 0x1000)? '0' : '') + code.toString(16);
  });
};

module.exports = smtpapi;
}());


examples/example.js
var smtpapi = require('../lib/main.js')
var nodemailer = require('nodemailer');
var header    = new smtpapi();

// addTo
header.addTo('test1@example.com');
header.addTo('test2@example.com');

// setTos
//header.setTos(['you@youremail.com', 'other@otheremail.com');

// addSubstitution
header.addSubstitution('keep', ['secret']);
header.addSubstitution('other', ['one', 'two']);

// setSubstitutions
//header.setSubstitutions({'keep': ['secret'], 'other': ['one', 'two']});

// addUniqueArg
header.addUniqueArg('cat', 'dogs');

// setUniqueArgs
//header.setUniqueArgs({cow: 'chicken'});
//header.setUniqueArgs({dad: 'proud'});

// addCategory
header.addCategory('tactics');
header.addCategory('advanced');

// setCategories
//header.setCategories(['tactics', 'advanced']);

// addSection
header.addSection('-charge-', 'This ship is useless.');

// setSections
//header.setSections({'-charge-': 'This ship is useless.', '-other': 'Another section here'});

// addFilter
// You can add filter settings one at a time.
header.addFilter('footer', 'enable', 1);
header.addFilter('footer', 'text/html', '<strong>boo</strong>');

// setFilters
// You can set a filter using an object literal.
/*
header.setFilters({
  'footer': {
    'setting': {
      'enable': 1,
      'text/plain': 'You can haz footers!'
    }
  }
});
*/

// setSendAt
// You can set the send_at scheduling param using an UNIX timestamp.
header.setSendAt(1409348513);

// setBatchId
// You can set the batch_id scheduling param using a string.
header.setBatchId('apples');

// setSendEachAt
// You can set the send_each_at scheduling param using a list of UNIX timestamps.
//header.setSendEachAt([1409348513, 14093485134]);

// addSendEachAt
// You can append one or more send_each_at scheduling param(s) to the existing ones using an UNIX timestamp.
//header.setSendEachAt([1409348511, 14093485132]);
//header.addSendEachAt(1409348513]);
//header.addSendEachAt(14093485134);

// setASMGroupID
// You can set the ASM Group ID.
header.setASMGroupID(123);

// [IP Pools](https://sendgrid.com/docs/API_Reference/Web_API_v3/IP_Management/ip_pools.html)
header.setIpPool("testPool");

// You can verify headers set (for debugging/logging purposes)
console.log(header.jsonString());

// Send usin Nodemailer
var headers    = { 'x-smtpapi': header.jsonString() };

// Use nodemailer to send the email
var settings  = {
  host: "smtp.sendgrid.net",
  port: parseInt(587, 10),
  requiresAuth: true,
  auth: {
    user: process.env.SENDGRID_USERNAME,
    pass: process.env.SENDGRID_PASSWORD
  }
};
var smtpTransport = nodemailer.createTransport(settings);

var mailOptions = {
  from:     "Example User <test@examplecom>",
  to:       "test@example.com",
  subject:  "Hello",
  text:     "Hello world",
  html:     "<b>Hello world</b>",
  headers:  headers
}

smtpTransport.sendMail(mailOptions, function(error, response) {
  smtpTransport.close();

  if (error) {
    console.log(error);
  } else {
    console.log("Message sent");
  }
});
